numba/cuda/cudaimpl.py

```diff
@@ -414,6 +414,50 @@ def ptx_selp(context, builder, sig, args):
     return builder.select(test, a, b)
 
 
+@lower(max, types.f4, types.f4)
+def ptx_max_f4(context, builder, sig, args):
+    fn = builder.module.get_or_insert_function(
+        lc.Type.function(
+            lc.Type.float(),
+            (lc.Type.float(), lc.Type.float())),
+        '__nv_fmaxf')
+    return builder.call(fn, args)
+
+
+@lower(max, types.f8, types.f4)
+@lower(max, types.f4, types.f8)
+@lower(max, types.f8, types.f8)
+def ptx_max_f8(context, builder, sig, args):
+    fn = builder.module.get_or_insert_function(
+        lc.Type.function(
+            lc.Type.double(),
+            (lc.Type.double(), lc.Type.double())),
+        '__nv_fmax')
+    return builder.call(fn, args)
```

Comment:
It isn't obvious to me why this implementation will work for the mixed type `(f4, f8)` and `(f8, f4)` cases without an explicit cast here.  The unit tests pass because this mixed type case is not checked.  Can you add a test and verify this works?

Reply:
in fact, I think if you leave off the mixed case type signature, that will force the upcast from `f4->f8` outside the function call, and that would be preferable.  Still need the mixed case test though to confirm.

Reply:
ok, stuart did some experimenting, and it seems that the mixed type signature is required to get this function to match, but I still suspect an explicit cast of both arguments to double is required before the `builder.call`